File: 06--wait-for-it/b.hs
#!/usr/bin/env runhaskell
-- Usage:
--   ./a.hs PATH_TO_INPUT_FILE
--   ./b.hs PATH_TO_INPUT_FILE

import Data.Function ((&))
import System.Environment (getArgs)


main = do
  [inputPath] <- getArgs
  text <- readFile inputPath
  putStrLn "Warning: If you're running b.hs on the puzzle input with runhaskell (interpreter), that will be slow. (Use `make build-haskell` and `make run-haskell` instead.)"
  print $ solve text


solve text =
  countWins time distance
  where
    [timesText, distancesText] = lines text

    parse line =
      line
      & words
      & drop 1
      & concat
      & readInt

    time = parse timesText
    distance = parse distancesText


countWins time recordDistance =
  [0..time]
  & map
      (\chargeTime ->
        let
          moveTime = time - chargeTime
          speed = chargeTime
        in
          speed * moveTime
      )
  & filter
      (\distance ->
        distance > recordDistance
      )
  & length


readInt = read :: String -> Int

File: 06--wait-for-it/b.py
#!/usr/bin/env python3
'''
Usage:
    ./b.py PATH_TO_INPUT_FILE
'''

import sys
import math


def main():
    # Parse
    time_limit, record = open(sys.argv[1]).read().splitlines()
    time_limit = int(''.join(time_limit.split()[1:]))
    record = int(''.join(record.split()[1:]))

    # Simulate
    wins = 0
    for charge_time in range(0, time_limit+1):
        speed = charge_time
        move_time = time_limit - charge_time
        distance = speed * move_time
        if distance > record:
            wins += 1

    # Multiply
    print(wins)


if __name__ == '__main__':
    main()
